README.md
# mmlasso
This package provides the functions to calculate the MM-Lasso and adaptive MM-Lasso estimators proposed in Smucler and Yohai (2015).

## Development

If you add new code to `src/fnsauxarma.cpp` you will need to do the following two steps before installing the package.

```
library(Rcpp)
library(RcppArmadillo)
compileAttributes()
```


NEWS
Changes in mmlasso version 1.3.4

    + Some bugfixes in the C++ code.

    + Fixed a bug in the default parameters for sridge.

    + Default behavior for mmlasso and sridge now is to use only one core.

    + Fixed errata in the sridge help page.




R/mmlasso.R
mmlasso<-function(x,y,varsigma=1,cualcv.mm=5,cualcv.S=5,numlam.mm=30,numlam.S=30,niter.S=50,niter.mm=50,ncores=1){
  #Main function to compute MM-Lasso estimates.
  #The initial estimate is the S-Ridge of Maronna (2012).
  #INPUT
  #X, y data set
  #cualcv.mm: method for estimating prediction error of MM and adaptive MM-Lasso, cualcv-fold CV
  #cualcv.S: method for estimating prediction error of S-Ridge, cualcv-fold CV
  #numlam.mm: number of candidate lambda values for MM and adaptive MM-Lasso
  #numlam.S: number of candidate lambda values for S-Ridge
  #niter.mm : number of maximum iterations of weighted Lasso iterations for MM and adaptive MM-Lasso
  #niter.S : number of maximum iterations of IWLS for S-Ridge
  #ncores: number of cores to use for parallel computations. Default is one core.
  #varsigma: power to elevate the weights for the adaptive MM-Lasso
  #OUTPUT
  #coef.SE: Initial S-Ridge (p+1)-vector of estimated regression parameters, beta(1)=intercept
  #coef.MMLasso: Final MM-Lasso (p+1)-vector of estimated regression parameters, beta(1)=intercept
  #coef.MMLasso.ad: Final adaptive MM-Lasso (p+1)-vector of estimated regression parameters, beta(1)=intercept

  n<-nrow(x)
  p<-ncol(x)

  ###Center and scale covariates and response using median and mad
  prep<-prepara(x,y)
  xnor<-prep$Xnor
  ynor<-prep$ynor
  mux<-prep$mux
  sigx<-prep$sigx
  muy<-prep$muy
  ###

  ###Set-up cluster for parallel computations
  cores<-min(detectCores(),ncores)
  try(cl<-makeCluster(cores))
  try(registerDoParallel(cl))
  locallib<-.libPaths()[[1]]
  parallel::clusterExport(cl, "locallib", envir=environment())
  parallel::clusterEvalQ(cl,.libPaths(locallib))
  ###

  ###Calculate initial estimate and scale
  fit.SE<-sridge(xnor,ynor,normin=0,denormout=0,cualcv.S=cualcv.S,numlam.S=numlam.S,niter.S=niter.S) #Input data is already normalized, output is in normalized data scale
  beta.SE<-fit.SE$coef
  edf.SE<-fit.SE$edf+1
  beta.SE.slo<-beta.SE[2:length(beta.SE)]
  beta.SE.int<-beta.SE[1]
  scale.SE<-fit.SE$scale
  ###

  ###Adjust c1 as recommended by Maronna and Yohai (2010)
  if (edf.SE/n>0.1){
    c1<-3.89 #.90 de efi
  } else{
    c1<-3.44 #.85 de efi
  }
  ###

  ###Calculate candidate lambdas for MMLasso
  lambdamax0<-lambda0(xnor,ynor) #Initial candidate
  lambdamax1<-try(optim.lam(xnor,ynor,beta.SE,scale.SE,lambdamax0,c1,niter.mm))#Improvement
  if(class(lambdamax1)=='try-error'){
    lambdamax1<-lambdamax0
    warning('Using approximate lambdamax for MM-Lasso')
  }
  lambdas<-seq(0,lambdamax1,length.out=numlam.mm)
  if(p>=n){
    lambdas<-lambdas[2:numlam.mm]
  }
  ###

  ####Random permutation and order data for CV
  srt<-sort.int(sample(1:n),index.return=TRUE)
  tt<-srt$x
  orden<-srt$ix
  xnord<-xnor[orden,]
  ynord<-ynor[orden]
  ###

  ###Parallel CV
  exp.mm<-c('MMLasso','CVLasso')
  klam<-NULL
  mse<-foreach(klam=1:length(lambdas),.combine=c,.packages=c('mmlasso','robustHD'),.export=exp.mm)%dopar%{
    CVLasso(xnord,ynord,beta.SE,scale.SE,nfold=cualcv.mm,lambdas[klam],c1,niter.mm)
  }
  if(any(is.infinite(mse))){
    warning('Iteratively weighted Lasso failed when lambda equaled:')
    print(lambdas[which(is.infinite(mse))])
  }
  indmin<-which.min(mse)
  lamin<-lambdas[indmin]
  ###

  #Calculated final MMLasso estimate
  fit.MMLasso<-try(MMLasso(xnor,ynor,beta.SE,scale.SE,lamin,c1,niter.mm))
  if(class(fit.MMLasso)=='try-error'){
    beta.MMLasso.slo<-beta.SE[2:length(beta.SE)]
    beta.MMLasso.int<-beta.SE[1]
    warning('Calculation of the final MM-Lasso estimate failed, returning S-Ridge instead')
  }else{
  beta.MMLasso<-fit.MMLasso$coef
  beta.MMLasso.slo<-beta.MMLasso[2:length(beta.MMLasso)]
  beta.MMLasso.int<-beta.MMLasso[1]}

  activ <- which(beta.MMLasso.slo!=0)
  #If no variables were selected by the MMLasso, return the MMLasso
  if(length(activ)<=1){
    beta.MMLasso.slo<-beta.MMLasso.slo/sigx
    beta.MMLasso.int<-muy+beta.MMLasso.int-mux%*%beta.MMLasso.slo
    beta.SE.slo<-beta.SE.slo/sigx
    beta.SE.int<-muy+beta.SE.int-mux%*%beta.SE.slo
    beta.SE<-c(beta.SE.int,beta.SE.slo)
    beta.MMLasso<-c(beta.MMLasso.int,beta.MMLasso.slo)
    beta.MMLasso.ad<-beta.MMLasso
    re<-list(coef.MMLasso.ad=beta.MMLasso.ad,coef.MMLasso=beta.MMLasso,coef.SE=beta.SE)
    return(re)
  }


  ###Take out covariates not chosen by MMLasso and scale the remaining ones
  w.ad <- (abs(beta.MMLasso.slo[activ]))^varsigma
  xnor.w = as.matrix(scale(xnor[,activ],center=FALSE, scale = 1/w.ad))
  xnord.w <- as.matrix(xnor.w[orden,])
  beta.SE.w <- c(beta.SE[1],beta.SE[activ+1]/w.ad)

  ###Calculate candidate lambdas for adaptive MMLasso
  lambdamax0.ad<-lambda0(xnor.w,ynor)
  lambdamax1.ad<-try(optim.lam(xnor.w,ynor,beta.SE.w,scale.SE,lambdamax0.ad,c1,niter.mm))
  if (class(lambdamax1.ad)=='try-error'){
    lambdamax1.ad<-lambdamax0.ad
    warning('Using approximate lambdamax for adaptive MM-Lasso')
  }
  lambdas.ad<-seq(0,lambdamax1.ad,length.out=numlam.mm)
    if (p>=n){
      lambdas.ad<-lambdas.ad[2:numlam.mm]
    }
  ##

  ###Parallel CV for the adaptive MMLasso
  mse.ad<-foreach(klam=1:length(lambdas.ad),.combine=c,.packages=c('mmlasso','robustHD'),.export=exp.mm)%dopar%{
    CVLasso(xnord.w,ynord,beta.SE.w,scale.SE,nfold=cualcv.mm,lambdas.ad[klam],c1,niter.mm)
  }
  if(any(is.infinite(mse.ad))){
    warning('Iteratively weighted Lasso for adaptative MM-Lasso failed when lambda equaled:')
    print(lambdas.ad[which(is.infinite(mse.ad))])
  }
  indmin.ad<-which.min(mse.ad)
  lamin.ad<-lambdas.ad[indmin.ad]
  ###

  try(stopCluster(cl))

  ###Calculate final estimates and return to original coordinates

  fit.MMLasso.ad<-try(MMLasso(xnor.w,ynor,beta.SE.w,scale.SE,lamin.ad,c1,niter.mm)$coef)
  if(class(fit.MMLasso.ad)=='try-error'){
    beta.MMLasso.slo.ad<-beta.MMLasso.slo
    beta.MMLasso.int.ad<-beta.MMLasso.int
    warning('Calculation of the final adaptive MM-Lasso estimate failed, returning MM-Lasso instead')
  } else{
    beta.MMLasso.slo.ad <- rep(0,p)
    beta.MMLasso.slo.ad[activ] <- fit.MMLasso.ad[2:length(fit.MMLasso.ad)]*w.ad
    beta.MMLasso.int.ad <- fit.MMLasso.ad[1]
  }

  beta.MMLasso.slo.ad<-beta.MMLasso.slo.ad/sigx
  beta.MMLasso.int.ad<-muy+beta.MMLasso.int.ad-mux%*%beta.MMLasso.slo.ad
  beta.MMLasso.ad <- c(beta.MMLasso.int.ad,beta.MMLasso.slo.ad)

  beta.MMLasso.slo<-beta.MMLasso.slo/sigx
  beta.MMLasso.int<-muy+beta.MMLasso.int-mux%*%beta.MMLasso.slo
  beta.MMLasso<-c(beta.MMLasso.int,beta.MMLasso.slo)

  beta.SE.slo<-beta.SE.slo/sigx
  beta.SE.int<-muy+beta.SE.int-mux%*%beta.SE.slo
  beta.SE<-c(beta.SE.int,beta.SE.slo)

  ###
  re<-list(coef.MMLasso.ad=beta.MMLasso.ad,coef.MMLasso=beta.MMLasso,coef.SE=beta.SE)

  return(re)

}


CVLasso<-function(X,y,beta.ini,scale.ini,nfold,lam,c1,niter.mm){
  #Performs nfold-CV for MM-Lasso
  #INPUT
  #X,y: data
  #beta.ini, scale.ini: initial estimate of regression and scale
  #nfold: nfold-CV
  #lam: penalization parameter
  #c1: tuning constant for the rho function
  #niter.mm: maximum number of weighted Lasso iterations for MM-Lasso

  #OUTPUT
  #mse: resulting MSE (estimated using a tau-scale)

  ###Segment data
  n<-nrow(X)
  p<-ncol(X)
  indin<-1:n
  nestim<-n*(1-1/nfold) # Actual number of observations in an estim sample
  lamcv<-lam
  inint<-floor(seq(0,n,length.out=nfold+1))
  resid<-vector(mode='numeric',length=n)
  ###

  for (kk in 1:nfold){
    ###Get test and estimating samples
    testk<-(inint[kk]+1):inint[kk+1]
    estik<-setdiff(indin,testk);
    Xtest<-X[testk,]
    Xesti<-X[estik,]
    ytest<-y[testk]
    yesti<-y[estik]
    fit.MMBR<-try(MMLasso(Xesti,yesti,beta.ini,scale.ini,lambda=lamcv,c1=c1,niter.mm))
    if (class(fit.MMBR)=="try-error"){
      return(Inf)
    }
    beta.MMBR<-fit.MMBR$coef
    beta.MMBR.slo<-beta.MMBR[2:length(beta.MMBR)]
    beta.MMBR.int<-beta.MMBR[1]
    fitk<-Xtest%*%beta.MMBR.slo+beta.MMBR.int
    resid[testk]<-ytest-fitk
  }
  mse<-scale_tau(resid)
  return(mse)
}


MMLasso<-function(xx,y,beta.ini,scale.ini,lambda,c1,niter.mm){
  #Performs iteratively re-weighted Lasso
  #INPUT
  #xx,yy: data
  #beta.ini, scale.ini: initial estimate of regression and scale
  #lambda: penalization parameter
  #c1: tuning constant for the rho function
  #niter.mm: maximum number of iterations
  #OUTPUT:
  #coef: final estimate

  MMcpp_ini<-MMLassoCpp_ini(xx,y,beta.ini)
  x<-MMcpp_ini$x
  resid.n<-MMcpp_ini$resid.n
  tol<-1
  m<-0
  beta.n<-beta.ini
  p<-length(beta.ini)

  while (tol>= 1e-4){
    beta.o<-beta.n
    if(all(beta.o==0)){
      return(list(coef=beta.o))
    }
    MMcpp1<-MMLassoCpp1(x,y,beta.o,scale.ini,c1)
    xort<-MMcpp1$xort
    xjota<-MMcpp1$xjota
    yast<-MMcpp1$yast
    alpha<-MMcpp1$alpha
    u.Gram<- !(p>=500)
    if(all(xjota==0)){
      return(list(coef=beta.o))
    }
    try(res.Lasso <- robustHD:::fastLasso(xort, yast,lambda,intercept=FALSE, normalize=FALSE, use.Gram=u.Gram),silent=TRUE)
    if (class(res.Lasso)=="try-error"){
      warning("fastLasso failed")
      return(list(coef=beta.o))
    }
    beta.Lasso <- coef(res.Lasso)
    MMcpp2<-MMLassoCpp2(xjota,yast,beta.Lasso,beta.o,alpha)
    beta.n<-MMcpp2$beta.n
    tol<-MMcpp2$tol
    m<-m+1
    if (m >= niter.mm) {tol<-0}
  }
  return(list(coef=beta.n))
}



optim.lam<-function(x,y,beta.ini,scale.ini,lambdamax,c1,niter.mm){
  #Finds (approximately) smallest lambda such that the slope of the MM-Lasso estimate is zero
  #INPUT
  #X, y: data set
  #beta.ini, scale.ini: initial estimates and scale
  #lambdamax: initial guess for lambda
  #c1: tuning constant for the rho function
  #niter.mm: maximum number of iterations
  #OUTPUT
  #lambda: smallest lambda such that the slope of the MM-Lasso estimate is zero

  n<-nrow(x)
  p<-ncol(x)
  lambda2<-lambdamax
  beta.n<-MMLasso(x,y,beta.ini,scale.ini,lambda2,c1,niter.mm)$coef
  zeros.n<-sum(beta.n[2:(p+1)]==0)
  while (zeros.n<p & lambda2<max(n,1e4)){
    lambda2<-2*lambda2
    beta.n<-MMLasso(x,y,beta.ini,scale.ini,lambda2,c1,niter.mm)$coef
    zeros.n<-sum(beta.n[2:(p+1)]==0)
  }
  if(lambda2>=max(n,1e4))
  {
    return(lambda2)
  }
  lambda1<-lambdamax/2
  beta.o<-MMLasso(x,y,beta.ini,scale.ini,lambda1,c1,niter.mm)$coef
  zeros.o<-sum(beta.o[2:(p+1)]==0)
  while(zeros.o==p){
    lambda1<-lambda1/2
    beta.o<-MMLasso(x,y,beta.ini,scale.ini,lambda1,c1,niter.mm)$coef
    zeros.o<-sum(beta.o[2:(p+1)]==0)
  }
  for (j in 1:5){
    lambda<-0.5*(lambda1+lambda2)
    beta.n<-MMLasso(x,y,beta.ini,scale.ini,lambda,c1,niter.mm)$coef
    zeros.n<-sum(beta.n[2:(p+1)]==0)
    if (zeros.n<p){
      lambda1<-lambda
    }else{
      lambda2<-lambda
    }
  }
  return(lambda2)
}



prepara<-function(X,y){
  #Centers y and the columns of X to zero median, and normalizes X to unit mad
  #INPUT
  #X, y: data set
  #OUTPUT
  #Xnor: centered normalized X
  #ycen: centered y
  #mux: median vector of X
  #scalex: mad vector of X
  #muy: median of y
  #sigy:scale of y
  muy<-median(y)
  ycen<-y-muy
  sigy<-mad(y)
  mux<-apply(X,2,'median')
  sigx<-apply(X,2,'mad')
  Xnor<-scale(X,center=mux,scale=sigx)
  res<-list(Xnor=Xnor,ynor=ycen,mux=mux,sigx=sigx,muy=muy,sigy=sigy)
  return(res)
}


R/sridge.R
sridge <- function(x, y, cualcv.S=5, numlam.S=30, niter.S=50, normin=0, denormout=0, alone=0, ncores=1){
  # Solves n*s_n^2 +lam*||beta1||^2} = min. Adapted from Ricardo Maronna's original MATLAB code.
  # INPUT
  # cualcv.S: method for estimating prediction error. cualcv-fold cross-validation
  # normin: normalize input data?. 0=no,  default ; 1=yes
  # denormout: denormalize output?. 0=no,  default ; 1=yes
  # alone: are you calculating the estimator for its sake only? 0=no,  default ; 1=yes
  # numlam.S: number of lambda values,  default 30
  # niter.S : number of maximum iterations of IWLS
  # ncores : number of cores to use for parallel computations. Default is one core.
  # OUTPUT
  # coef: (p+1)-vector of regression parameters,  beta(1)=intercept
  # scale: M-estimate of scale of the final regression estimate
  # edf: final equivalent degrees of freedom
  # lamda: optimal lambda
  # delta: optimal delta


  ###Center and scale covariates and response using median and mad
  if (normin == 1){
    prep <- prepara(x,  y)
    Xnor <- prep$Xnor
    ynor <- prep$ynor
    mux <- prep$mux
    sigx <- prep$sigx
    muy <- prep$muy
  }else{
    Xnor <- x
    ynor <- y
  }

  # Spherical Principal Components (no centering) privar,  Beig= vector of robust
  # "eigenvalues" and matrix of eigenvectors Xnor is now = PCA scores
  pca <- SPCC(Xnor)
  privar <- pca$lamda
  Beig <- pca$b
  Xnor <- pca$scores
  n <- nrow(Xnor)
  p <- ncol(Xnor)
  nkeep <- 5 # number of candidates to be kept for full iteration in the Pena-Yohai procedure
  privar <- privar*n # Makes the robust eigenvalues of the same order as those of classical PCA used for LS
  nlam <- min(c(p, numlam.S))
  pmax <- min(c(p,  n/2))   # edf<=n/2 to keep BDP >=0.25
  pp <- seq(1, pmax, length=nlam)
  lamdas <- findlam(privar, pp) # find lambdas corresponding to the edf's
  deltas <- 0.5*(1-(pp)/n)  # for the M-scale used with Penia-Yohai

  ### Reorder data for CV
  srt <- sort.int(sample(1:n), index.return=TRUE)  # Random permutation
  tt <- srt$x
  orden <- srt$ix
  Xnord <- Xnor[orden, ]
  ynord <- ynor[orden]
  ###

  if (alone == 1){
    if(ncores > 1){
      if(length(find("cl")) != 1){
	newclus <- TRUE
	### Set-up cluster for parallel computations
	cores <- min(detectCores(), ncores)
	try(cl <- makeCluster(cores))
      } else {
	newclus <- FALSE
	try(registerDoParallel(cl))
	locallib <- .libPaths()[[1]]
	parallel::clusterExport(cl,  "locallib",  envir=environment())
	parallel::clusterEvalQ(cl, .libPaths(locallib))
      }
    } else {
      newclus <- FALSE
    }
  }
  if(alone == 0){
    newclus <- length(find("cl") != 1)
  }

  ### Cross Validation
  exp.se <- c("CVSE")
  klam <- NULL
  if(ncores == 1){
    mse <- foreach(klam = 1:length(lamdas),  .combine = c ) %do% {
      CVSE(Xnord, ynord, nfold=cualcv.S, lamdas[klam], pp[klam], nkeep, niter.S)
    }
  } else {
    mse <- foreach(klam = 1:length(lamdas),  .combine = c,  .packages = c("mmlasso"),
		   .export = exp.se) %dopar% {
      CVSE(Xnord, ynord, nfold=cualcv.S, lamdas[klam], pp[klam], nkeep, niter.S)
    }
  }

  if(any(is.infinite(mse))){
    warning("IWLS for S-Ridge failed when lambda equaled:")
    print(lamdas[which(is.infinite(mse))])
  }
  indmin <- which.min(mse)
  lamin <- lamdas[indmin]
  delmin <- deltas[indmin]
  ###

  if (alone == 1 & newclus){
    stopCluster(cl)
  }

  fin <- rr_se(Xnor, ynor, lamin, deltaesc=delmin, cc_scale=1, nkeep, niter.S, epsilon=1e-4)
  beta <- fin$coef
  betaslo <- beta[2:(length(beta))]
  bint <- beta[1]
  res <- ynor-Xnor%*%betaslo-as.vector(bint)
  edf <- fin$edf
  deltult <- 0.5 * (1 - (edf + 1)/n)  ##  'delta' for final scale
  deltult <- max(c(deltult,  0.25))
  # c0: constant for consistency of final scale
  c0 <- const_marina(deltult)
  sigma <- Mscale_mar(res, deltult, c0)
  a_cor <- mean(psi_marina(res/sigma, c0)^2)
  b_cor <- mean(Mchi(res/sigma,  c0,  "bisquare",  2))
  c_cor <- mean(psi_marina(res/sigma, c0)*(res/sigma))
  corr <- 1+(edf+1)/(2*n)*(a_cor/(b_cor*c_cor))
  fscale <- sigma*corr  # bias correction for final scale
  # Back from PC to ordinary coordinates
  betaslo <- Beig%*%betaslo

  # De-normalize beta if required by user
  if (normin==1 & denormout==1){
    betaslo <- betaslo/sigx
    bint <- muy+bint-mux%*%betaslo
  }
  beta <- c(bint, betaslo)
  re <- list(coef=beta, scale=fscale, edf=edf, lamda=lamin, delta=delmin)
  return(re)
}


CVSE <- function(X, y, nfold, lam, gradlib, nkeep, niter.S){
  # Performs nfold-CV for S-Ridge
  # INPUT
  # beta.ini,  scale.ini: initial estimates of regression and scale
  # X, y: data
  # lam: penalization parameter
  # gradlib: degrees of freedom,  used to calculate the delta for the M-scale
  # nkeep: number of candidates for full iterations of IWLS
  # niter.S : number of maximum iterations of IWLS
  # OUTPUT
  # mse: resulting MSE (estimated using a tau-scale)

  ### Segment data
  n <- nrow(X)
  p <- ncol(X)
  indin <- 1:n
  nestim <- n*(1-1/nfold)
  lamcv <- lam
  deltaesc <- 0.5*(1-(gradlib)/nestim)
  inint <- floor(seq(0, n, length.out=nfold+1))
  resid <- vector(mode = "numeric",  length = n)
  ###

  for (kk in 1:nfold){
    testk <- (inint[kk]+1):inint[kk+1]
    estik <- setdiff(indin,  testk)
    Xtest <- X[testk, ]
    Xesti <- X[estik, ]
    ytest <- y[testk]
    yesti <- y[estik]
    se <- try(rr_se(Xesti, yesti, lamcv, deltaesc, cc_scale=1, nkeep, niter.S, epsilon=1e-4))
    if (class(se)=="try-error"){
      return(Inf)
    }
    bet <- se$coef
    bint <- bet[1]
    beta <- bet[2:(p+1)]
    fitk <- Xtest%*%beta+bint
    resid[testk] <- ytest-fitk
  }
  mse <- scale_tau(resid)
  return(mse)
}

findlam <- function(vals, r){
  # Finds lamdas which yield edf=r
  p <- length(vals)
  nr <- length(r)
  lamr <- vector(mode = "numeric",  nr)
  lam1 <- 0
  lam2 <- max(vals)*(p/r-1)
  for (i in 1:nr){
    lam0 <- c(lam1,  lam2[i]+0.5)   # the value 0.5 is for lam=0
    lamr[i] <- uniroot(sumrat, lam0, vals, r[i])$root
  }
  return(lamr)
}

sumrat <- function(lam, vals, r){
  susu <- sum(vals/(vals+lam))-r
  return(susu)
}


const_marina <- function(delta){
  integrand <-  function(x, c){dnorm(x)*rho_marina(x, c)}
  expe <- function(c, delta){integrate(integrand, lower=-Inf, upper=Inf, c)$value-delta}
  init <- c(0.1, 100)
  try(cw <- uniroot(expe, init, delta)$root, silent=TRUE)
  if (class(cw)=="try-error"){
    warning("Something's wrong,  could not find tuning constant for the scale")
    return(NULL)
  }
  return(cw)
}


#' A function to fit a ridge penalized MM-estimator. Solves n*s_n^2 +lam*||beta1||^2 = min. Adapted from Ricardo Maronna's original
#'  MATLAB code.
#' This version does not do any cross-validation. It merely returns the fits at the chosen lambdas.
#' @param x a matrix of covariates (without intercept)
#' @param y a vector of response values
#' @param lamdas are a vector of lambda values or a scalar lambda value.
#' @param deltas should be the delta values associated with the lambda values.
#' @param nkeep: number of candidates to be kept for full iteration in the Pena-Yohai procedure (default=5)
#' @param normin: normalize input  data?. 0=no,  default ; 1=yes
#' @param denormout: denormalize output?. 0=no,  default ;  1=yes
#' @param alone: are you calculating the estimator for its sake only? 0=no,  default  ; 1=yes
#' @param numlam.S: number of lambda values,  default 30
#' @param niter.S : number of maximum iterations of IWLS
#' @param ncores : number of cores to use for parallel computations. Default is 1 (i.e. no parallelization)
#' @return a list of  beta: (p+1)-vector of  regression parameters,  beta(1)=intercept fscale: M-estimate of scale of the final regression estimate edf: final equivalent degrees of freedom.
#' @export
sridge2 <- function(x,  y, nkeep = 5,  numlam.S = 30,  niter.S = 50,  normin = 0,
		    denormout = 0,  alone = 0,  ncores = 1,  lamdas = NULL,  deltas = NULL,
		    lamdasad = NULL, deltasad = NULL, adaptive = FALSE) {

  ###  Center and scale covariates and response using median and mad
  if (normin == 1) {
    stop("Not able to scale by mad when most of variables are binary. Maybe will scale in some other way later.")
    prep <- prepara(x,  y)
    Xnor <- prep$Xnor
    ynor <- prep$ynor
    mux <- prep$mux
    sigx <- prep$sigx
    muy <- prep$muy
  } else {
    Xnor <- x
    ynor <- y
    sigx <- rep(1, ncol(x))
  }

  #  Spherical Principal Components (no centering) privar,  Beig= vector of robust
  #  "eigenvalues" and matrix of eigenvectors Xnor is now = PCA scores
  pca <- SPCC(Xnor)
  privar <- pca$lamda
  Beig <- pca$b
  Xnor <- pca$scores
  n <- nrow(Xnor)
  p <- ncol(Xnor)
  if(is.null(lamdas)){ ##  Create lambda vector if one was not provided
    privar <- privar * n  ##  Makes the robust eigenvalues of the same order as those of classical PCA used for LS
    nlam <- min(c(p,  numlam.S))
    pmax <- min(c(p,  n/2))  # edf<=n/2 to keep BDP >=0.25
    pp <- seq(1,  pmax,  length = nlam)
    lamdas <- findlam(privar,  pp)  # find lambdas corresponding to the edf's
    deltas <- 0.5 * (1 - (pp)/n)  # for the M-scale used with Penia-Yohai
  }

  fin <- rr_se_vec(X = Xnor,  y = ynor,  lambda2 = lamdas,
		   deltaesc = deltas,  cc_scale = 1,  nkeep,  niter.S,  epsilon = 1e-04)
  betas <- fin$coef
  res <- apply(betas,2,function(b){
		 ynor - Xnor %*% b[-1] - as.vector(b[1])
		   })
  edf <- fin$edf
  deltult <- 0.5 * (1 - (edf + 1)/n)  ##  'delta' for final scale
  ## deltault should be no greater than .25
  ## deltult <- max(c(deltult,  0.25))
  ## deltult <- pmax(deltult,  rep(0.25,length(deltult)))
  deltult <- deltult * (deltult > .25) + .25 * ( 1 - (deltult > .25) )
  getfscale<-function(thedeltult,theres,theedf){
    #  c0: constant for consistency of final scale
    c0 <- const_marina(thedeltult)
    sigma <- Mscale_mar(theres,  thedeltult,  c0)
    a_cor <- mean(psi_marina(theres/sigma,  c0)^2)
    b_cor <- mean(Mchi(theres/sigma,  c0,  "bisquare",  2))
    c_cor <- mean(psi_marina(theres/sigma,  c0) * (theres/sigma))
    corr <- 1 + (theedf + 1)/(2 * n) * (a_cor/(b_cor * c_cor))
    fscale <- sigma * corr  # bias correction for final scale
    return(fscale)
  }
  fscale <- sapply(1:ncol(res),function(i){
		     getfscale(thedeltult=deltult[i],
			       theres=res[,i],theedf=edf[i]) })
  #  Back from PC to ordinary coordinates
  betaslo <- apply(betas[-1,],2,function(b){ Beig %*% b})

  if(!adaptive){
    betares <- rbind(betas[1,],  betaslo)
    results <- list(coef = betares,  scale = fscale,  edf = edf,  lamda = lamdas,  delta = deltas,
		    lamdasad = lamdasad, deltasad = deltasad)
    return(results)
  } else {
    resmaker <- function( Xnor,  ynor,  nkeep,  niter.S,  Beigi, numlam.S){
      force(Xnor); force(ynor); force(nkeep); force(niter.S); force(Beig); force(numlam.S)
      function(lamdasad, deltasad, thebetas){
	## For each initial model (i.e. vector of betas in thebetas), use those betas as weights and fit another set of models.
	activ <- which(abs(thebetas[2:length(thebetas)]) > .Machine$double.eps)
	## Take out covariates not chosen by MMLasso and scale the remaining ones
	wad <- abs(thebetas[2:length(thebetas)][activ])
	Xnorw <- as.matrix(scale(Xnor[, activ], center=FALSE,  scale = 1/wad))
	## Calculate candidate lambdas for adaptive MM-ridge
	if(is.null(lamdasad) & is.null(deltasad)){
	  lamdasad <- genlambdas(Xnorw, ynor, numlam.S=numlam.S)
	  lamdasad <- lamdasad[lamdasad[,"lamda"]!=0,]
	  deltasad <- lamdasad[,"delta"]
	  lamdasad <- lamdasad[,"lamda"]
	}
	fin <- rr_se_vec(X = Xnorw,  y = ynor,  lambda2 = lamdasad,
			 deltaesc = deltasad,  cc_scale = 1,  nkeep,  niter.S,  epsilon = 1e-04)
	betas <- fin$coef
	betasbig <- matrix(0,ncol=length(thebetas),nrow=nrow(fin$coef))
	betasbig[,activ] <- fin$coef
	betasbig[-1,] <- betasbig[-1,]*wad
	res <- apply(betasbig,2,function(b){
		       ynor - Xnorw %*% b[-1] - as.vector(b[1])
			 })
	edfbig<-rep(0,p)
	edfbig[activ] <- fin$edf
	deltult <- 0.5 * (1 - (edfbig + 1)/n)  ##  'delta' for final scale
	deltult <- deltult * (deltult > .25) + .25 * ( 1 - (deltult > .25) )
	fscale <- sapply(1:ncol(res),function(i){
			   getfscale(thedeltult=deltult[i],
				     theres=res[,i],theedf=edf[i]) })
	#  Back from PC to ordinary coordinates
	betasload <- apply(betasbig[-1,],2,function(b){ Beig %*% b})
	beta <- rbind(betas[1,],  betasload)
	results <- list(coef = betasload,  scale = fscale,  edf = edf,
			lamdasad = lamdasad, deltasad = lamdasad)
	return(results)

	##  De-normalize beta if required by user
	##  if (normin == 1 & denormout == 1) {
	##      betaslo <- betaslo/sigx
	##      bint <- muy + bint - mux %*% betaslo
	##  }
      }
    }

    getresults <- resmaker(Xnor,  ynor,  nkeep,  niter.S,  Beig, numlam.S)

    if(ncores > 1){
      ##  assuming name of cluster is cl for now
      clusterExport(cl, c("getresults","getfscale"), envir=environment())
      results <- parLapply(cl, 1:ncol(betas),  function(i){
			     tmp <- getresults(lamdasad=lamdasad, deltasad=deltasad, thebetas=betas[,i])
			     tmp[["lamda"]] <- lamdas
			     tmp[["deltas"]] <- deltas
			     return(tmp)
			})
    } else {
      results <- lapply(1:ncol(betas), function(i){
			  message(i)
			  tmp <- getresults(lamdasad=lamdasad, deltasad=deltasad, thebetas=betas[,i])
			  tmp[["lamda"]] <- lamdas
			  tmp[["deltas"]] <- deltas
			  return(tmp)
			})
    }

    return(results)
  }
}

#' Generate lambda sequences
genlambdas <- function(x,  y,  normin = 0,  numlam.S = 30) {

  ###  Center and scale covariates and response using median and mad
  if (normin == 1) {
    stop("Not able to scale by mad when most of variables are binary. Maybe will scale in some other way later.")
    prep <- prepara(x,  y)
    Xnor <- prep$Xnor
    ynor <- prep$ynor
    mux <- prep$mux
    sigx <- prep$sigx
    muy <- prep$muy
  } else {
    Xnor <- x
    ynor <- y
  }

  #  Spherical Principal Components (no centering) privar,  Beig= vector of robust
  #  'eigenvalues' and matrix of eigenvectors Xnor is now = PCA scores
  pca <- SPCC(Xnor)
  privar <- pca$lamda
  Beig <- pca$b
  Xnor <- pca$scores
  n <- nrow(Xnor)
  p <- ncol(Xnor)
  privar <- privar * n  # Makes the robust eigenvalues of the same order as those of classical PCA used for LS
  nlam <- min(c(p,  numlam.S))
  pmax <- min(c(p,  n/2))  # edf<=n/2 to keep BDP >=0.25
  pp <- seq(1,  pmax,  length = nlam)
  lamdas <- findlam(privar,  pp)  # find lambdas corresponding to the edf's
  deltas <- 0.5 * (1 - (pp)/n)  # for the M-scale used with Penia-Yohai

  return(cbind(lamda=lamdas, delta=deltas))
}
